File: android/app/src/main/kotlin/br/com/tportooliveira/fotocelula/results/EventStore.kt
package br.com.tportooliveira.fotocelula.results

import android.content.Context
import br.com.tportooliveira.fotocelula.core.EventScoring
import br.com.tportooliveira.fotocelula.core.TimeFormatter
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.Locale
import java.util.UUID
import java.util.concurrent.Executors

/** Uma prova (evento). Vive só no aparelho: sem conta, sem servidor, sem rede. */
data class Event(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val dateMillis: Long = System.currentTimeMillis(),
    val place: String = "",
    val notes: String = "",
) {
    fun toJson(): JSONObject = JSONObject().apply {
        put("id", id); put("name", name); put("dateMillis", dateMillis); put("place", place); put("notes", notes)
    }

    companion object {
        fun fromJson(j: JSONObject) = Event(
            id = j.getString("id"), name = j.optString("name"), dateMillis = j.optLong("dateMillis"),
            place = j.optString("place"), notes = j.optString("notes"),
        )
    }
}

/** Uma inscrição na prova: quem larga, em que ordem, em que categoria. */
data class Entry(
    val id: String = UUID.randomUUID().toString(),
    val eventId: String,
    val order: Int,
    val rider: String,
    val horse: String = "",
    val category: String = "",
) {
    /** Rótulo curto para a faixa "Próximo" (o operador lê de longe). */
    val label: String get() = buildString {
        append("#").append(order).append(' ').append(rider)
        if (horse.isNotBlank()) append(" / ").append(horse)
        if (category.isNotBlank()) append(" — ").append(category)
    }

    fun toJson(): JSONObject = JSONObject().apply {
        put("id", id); put("eventId", eventId); put("order", order)
        put("rider", rider); put("horse", horse); put("category", category)
    }

    companion object {
        fun fromJson(j: JSONObject) = Entry(
            id = j.getString("id"), eventId = j.getString("eventId"), order = j.optInt("order"),
            rider = j.optString("rider"), horse = j.optString("horse"), category = j.optString("category"),
        )
    }
}

/**
 * Provas e inscrições em JSON no armazenamento interno, no mesmo padrão do [RunHistoryStore]:
 * escrita fora da main thread, gravação atômica (temporário + rename) e leitura tolerante a arquivo
 * corrompido (o app abre vazio em vez de morrer na arena).
 */
class EventStore(context: Context) {
    private val file = File(context.filesDir, "provas.json")
    private val io = Executors.newSingleThreadExecutor { r -> Thread(r, "events-io") }
    private val _events = ArrayList<Event>()
    private val _entries = ArrayList<Entry>()
    /** Prova aberta no painel (nulo = cronômetro avulso, como o app era antes). */
    var currentEventId: String? = null
        private set

    val events: List<Event> get() = _events

    init { load() }

    fun entriesOf(eventId: String): List<Entry> =
        _entries.filter { it.eventId == eventId }.sortedWith(compareBy({ it.order }, { it.rider }))

    fun event(id: String?): Event? = _events.firstOrNull { it.id == id }

    fun entry(id: String?): Entry? = _entries.firstOrNull { it.id == id }

    fun addEvent(e: Event) { _events.add(0, e); currentEventId = e.id; save() }

    fun updateEvent(e: Event) {
        val i = _events.indexOfFirst { it.id == e.id }
        if (i >= 0) { _events[i] = e; save() }
    }

    fun removeEvent(id: String) {
        _events.removeAll { it.id == id }
        _entries.removeAll { it.eventId == id }
        if (currentEventId == id) currentEventId = null
        save()
    }

    fun select(id: String?) { currentEventId = id; save() }

    fun addEntry(e: Entry) { _entries.add(e); save() }

    fun updateEntry(e: Entry) {
        val i = _entries.indexOfFirst { it.id == e.id }
        if (i >= 0) { _entries[i] = e; save() }
    }

    fun removeEntry(id: String) { _entries.removeAll { it.id == id }; save() }

    /**
     * Próxima inscrição a largar: a de menor ordem que ainda não tem passada salva. `records` é o
     * histórico completo (o vínculo é por `entryId`).
     */
    fun nextEntry(eventId: String, records: List<RunRecord>): Entry? {
        val done = records.mapNotNull { it.entryId }.toHashSet()
        return entriesOf(eventId).firstOrNull { it.id !in done }
    }

    /**
     * Importa a lista de largada de um CSV `ordem;competidor;cavalo;categoria` (separador ";" ou ",",
     * cabeçalho opcional). Devolve quantas inscrições entraram. Linhas inválidas são ignoradas — a
     * planilha vem de terceiros e não pode derrubar o app na hora da prova.
     */
    fun importEntries(eventId: String, csv: String): Int {
        var added = 0
        var nextOrder = (entriesOf(eventId).maxOfOrNull { it.order } ?: 0) + 1
        for (raw in csv.lineSequence()) {
            val line = raw.trim().removePrefix("﻿")
            if (line.isEmpty()) continue
            val sep = if (line.count { it == ';' } >= line.count { it == ',' }) ';' else ','
            val cols = splitCsvLine(line, sep)
            if (cols.isEmpty()) continue
            val order = cols[0].toIntOrNull()
            // cabeçalho ou linha sem número: o nome pode estar na primeira coluna
            val rider = if (order != null) cols.getOrElse(1) { "" } else cols[0]
            if (rider.isBlank()) continue
            if (order == null && rider.lowercase(Locale.ROOT).let {
                    it == "competidor" || it == "nome" || it == "ordem"
                }) continue
            val horse = if (order != null) cols.getOrElse(2) { "" } else cols.getOrElse(1) { "" }
            val category = if (order != null) cols.getOrElse(3) { "" } else cols.getOrElse(2) { "" }
            _entries.add(Entry(eventId = eventId, order = order ?: nextOrder, rider = rider,
                horse = horse, category = category))
            if (order == null) nextOrder++ else nextOrder = maxOf(nextOrder, order + 1)
            added++
        }
        if (added > 0) save()
        return added
    }

    /** Classificação da prova por categoria (regra do núcleo compartilhado). */
    fun ranking(eventId: String, records: List<RunRecord>): List<EventScoring.Placing> =
        EventScoring.rankByCategory(records.filter { it.eventId == eventId }.map { it.toScoringRun() })

    /** CSV da prova: colocação, competidor, tempos, penalidade — o que se imprime ou se manda no zap. */
    fun rankingCsv(eventId: String, records: List<RunRecord>): String {
        fun dec(v: Double, places: Int) = String.format(Locale.ROOT, "%.${places}f", v).replace('.', ',')
        fun q(s: String) = "\"" + s.replace("\"", "\"\"") + "\""
        val mine = records.filter { it.eventId == eventId }
        val byOrder = mine.associateBy { it.entryOrder }
        val sb = StringBuilder()
        sb.append("categoria;colocacao;ordem;competidor;cavalo;tempo_final;tempo_bruto_s;tambores;penalidade_s;sem_tempo\n")
        for (p in ranking(eventId, records)) {
            val r = byOrder[p.entryOrder] ?: continue
            val e = _entries.firstOrNull { it.id == r.entryId }
            sb.append(listOf(
                q(r.category), p.place?.toString() ?: "SAT", "${p.entryOrder}",
                q(e?.rider ?: r.rider), q(e?.horse ?: r.horse),
                if (r.noTime) "SAT" else TimeFormatter.formatElapsed(p.finalNs).replace('.', ','),
                dec(r.elapsedRawNs / 1e9, 3), "${r.barrelsKnocked}", "${p.penaltyNs / 1_000_000_000L}",
                if (r.noTime) "sim" else "não",
            ).joinToString(";")).append('\n')
        }
        return sb.toString()
    }

    // ------------------------------------------------------------------ persistência
    private fun load() {
        if (!file.exists()) return
        try {
            val root = JSONObject(file.readText())
            val ev = root.optJSONArray("events") ?: JSONArray()
            for (i in 0 until ev.length()) _events.add(Event.fromJson(ev.getJSONObject(i)))
            val en = root.optJSONArray("entries") ?: JSONArray()
            for (i in 0 until en.length()) _entries.add(Entry.fromJson(en.getJSONObject(i)))
            currentEventId = if (root.isNull("currentEventId")) null else root.optString("currentEventId").ifEmpty { null }
            if (event(currentEventId) == null) currentEventId = null
        } catch (_: Exception) {
            _events.clear(); _entries.clear(); currentEventId = null
        }
    }

    private fun save() {
        val root = JSONObject()
        val ev = JSONArray(); for (e in _events) ev.put(e.toJson())
        val en = JSONArray(); for (e in _entries) en.put(e.toJson())
        root.put("events", ev); root.put("entries", en)
        currentEventId?.let { root.put("currentEventId", it) }
        val text = root.toString(2)
        io.execute {
            try {
                val tmp = File(file.path + ".tmp")
                tmp.writeText(text)
                if (!tmp.renameTo(file)) file.writeText(text)
            } catch (_: Exception) { }
        }
    }

    /** Divide uma linha de CSV respeitando aspas duplas (planilhas exportam nomes com ponto e vírgula). */
    private fun splitCsvLine(line: String, sep: Char): List<String> {
        val out = ArrayList<String>()
        val cur = StringBuilder()
        var quoted = false
        var i = 0
        while (i < line.length) {
            val c = line[i]
            when {
                quoted && c == '"' && i + 1 < line.length && line[i + 1] == '"' -> { cur.append('"'); i++ }
                c == '"' -> quoted = !quoted
                c == sep && !quoted -> { out.add(cur.toString().trim()); cur.setLength(0) }
                else -> cur.append(c)
            }
            i++
        }
        out.add(cur.toString().trim())
        return out
    }
}

File: ios/App/Results/EventStore.swift
import Combine
import Foundation
import PhotocellCore

/// Uma linha da classificação: a colocação e a passada que a produziu.
/// Tipo nomeado, não tupla — `ForEach(_:id:)` precisa de um key path, e key path para elemento de
/// tupla não existe em Swift.
struct RankingRow: Identifiable {
    let placing: EventScoring.Placing
    let run: RunRecord
    var id: UUID { run.id }
}

/// Uma prova (evento). Vive só no aparelho: sem conta, sem servidor, sem rede.
struct Event: Codable, Identifiable, Equatable {
    var id: UUID = UUID()
    var name: String
    var date: Date = Date()
    var place: String = ""
    var notes: String = ""
}

/// Uma inscrição na prova: quem larga, em que ordem, em que categoria.
struct Entry: Codable, Identifiable, Equatable {
    var id: UUID = UUID()
    var eventId: UUID
    var order: Int
    var rider: String
    var horse: String = ""
    var category: String = ""

    /// Rótulo curto para a faixa "Próximo" (o operador lê de longe).
    var label: String {
        var s = "#\(order) \(rider)"
        if !horse.isEmpty { s += " / \(horse)" }
        if !category.isEmpty { s += " — \(category)" }
        return s
    }
}

/// Provas e inscrições em JSON dentro de Application Support, no mesmo padrão do `RunHistoryStore`:
/// gravação atômica e leitura tolerante a arquivo corrompido (o app abre vazio em vez de morrer na arena).
@MainActor
final class EventStore: ObservableObject {
    @Published private(set) var events: [Event] = []
    @Published private(set) var entries: [Entry] = []
    /// Prova aberta no painel (nulo = cronômetro avulso, como o app era antes).
    @Published private(set) var currentEventId: UUID?

    private let fileURL: URL

    private struct Payload: Codable {
        var events: [Event]
        var entries: [Entry]
        var currentEventId: UUID?
    }

    init() {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("FotocelulaTambor", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        fileURL = dir.appendingPathComponent("provas.json")
        load()
    }

    // MARK: - consultas
    var currentEvent: Event? { events.first { $0.id == currentEventId } }

    func entries(of eventId: UUID) -> [Entry] {
        entries.filter { $0.eventId == eventId }.sorted { ($0.order, $0.rider) < ($1.order, $1.rider) }
    }

    func entry(_ id: UUID?) -> Entry? {
        guard let id else { return nil }
        return entries.first { $0.id == id }
    }

    /// Próxima inscrição a largar: a de menor ordem que ainda não tem passada salva.
    func nextEntry(of eventId: UUID, records: [RunRecord]) -> Entry? {
        let done = Set(records.compactMap { $0.entryId })
        return entries(of: eventId).first { !done.contains($0.id) }
    }

    // MARK: - edição
    func add(_ e: Event) { events.insert(e, at: 0); currentEventId = e.id; save() }

    func update(_ e: Event) {
        if let i = events.firstIndex(where: { $0.id == e.id }) { events[i] = e; save() }
    }

    func remove(eventId: UUID) {
        events.removeAll { $0.id == eventId }
        entries.removeAll { $0.eventId == eventId }
        if currentEventId == eventId { currentEventId = nil }
        save()
    }

    func select(_ id: UUID?) { currentEventId = id; save() }

    func add(_ e: Entry) { entries.append(e); save() }

    func update(_ e: Entry) {
        if let i = entries.firstIndex(where: { $0.id == e.id }) { entries[i] = e; save() }
    }

    func remove(entryId: UUID) { entries.removeAll { $0.id == entryId }; save() }

    /// Importa a lista de largada de um CSV `ordem;competidor;cavalo;categoria` (separador ";" ou ",",
    /// cabeçalho opcional). Devolve quantas inscrições entraram. Linhas inválidas são ignoradas — a
    /// planilha vem de terceiros e não pode derrubar o app na hora da prova.
    @discardableResult
    func importEntries(into eventId: UUID, csv: String) -> Int {
        var added = 0
        var nextOrder = (entries(of: eventId).map(\.order).max() ?? 0) + 1
        for raw in csv.split(whereSeparator: \.isNewline) {
            let line = raw.trimmingCharacters(in: .whitespaces).replacingOccurrences(of: "\u{FEFF}", with: "")
            if line.isEmpty { continue }
            let sep: Character = line.filter { $0 == ";" }.count >= line.filter { $0 == "," }.count ? ";" : ","
            let cols = Self.splitCsvLine(line, separator: sep)
            if cols.isEmpty { continue }
            let order = Int(cols[0])
            let rider = order != nil ? (cols.count > 1 ? cols[1] : "") : cols[0]
            if rider.trimmingCharacters(in: .whitespaces).isEmpty { continue }
            if order == nil, ["competidor", "nome", "ordem"].contains(rider.lowercased()) { continue }
            let horse = order != nil ? (cols.count > 2 ? cols[2] : "") : (cols.count > 1 ? cols[1] : "")
            let category = order != nil ? (cols.count > 3 ? cols[3] : "") : (cols.count > 2 ? cols[2] : "")
            entries.append(Entry(eventId: eventId, order: order ?? nextOrder, rider: rider,
                                 horse: horse, category: category))
            if let order { nextOrder = max(nextOrder, order + 1) } else { nextOrder += 1 }
            added += 1
        }
        if added > 0 { save() }
        return added
    }

    // MARK: - classificação
    /// Classificação da prova por categoria (regra do núcleo compartilhado).
    func ranking(of eventId: UUID, records: [RunRecord]) -> [EventScoring.Placing] {
        rankingRows(of: eventId, records: records).map(\.placing)
    }

    /// Classificação já pareada com a passada de cada colocação.
    ///
    /// O pareamento é feito DENTRO de cada categoria: a ordem de largada costuma ser numerada por
    /// categoria (o #1 da Amador e o #1 da Aberta existem os dois), então um mapa por `entryOrder`
    /// sobre a prova inteira mostraria o competidor errado. Tela e CSV usam esta mesma função.
    func rankingRows(of eventId: UUID, records: [RunRecord]) -> [RankingRow] {
        let mine = records.filter { $0.eventId == eventId }
        var cats: [String] = []
        for r in mine where !cats.contains(r.category ?? "") { cats.append(r.category ?? "") }
        var out: [RankingRow] = []
        for cat in cats.sorted(by: { $0.unicodeScalars.lexicographicallyPrecedes($1.unicodeScalars) }) {
            let inCat = mine.filter { ($0.category ?? "") == cat }
            // dentro da categoria a ordem ainda pode repetir (duas passadas do mesmo número): a
            // colocação sai na mesma ordem do ranking, então consome-se a fila por posição
            var remaining: [Int: [RunRecord]] = [:]
            for r in inCat { remaining[r.entryOrder ?? 0, default: []].append(r) }
            for p in EventScoring.rank(inCat.map(\.scoringRun)) {
                guard var queue = remaining[p.entryOrder], !queue.isEmpty else { continue }
                let r = queue.removeFirst()
                remaining[p.entryOrder] = queue
                out.append(RankingRow(placing: p, run: r))
            }
        }
        return out
    }

    /// CSV da prova: colocação, competidor, tempos, penalidade — o que se imprime ou se manda no zap.
    func rankingCsv(of eventId: UUID, records: [RunRecord]) -> String {
        func dec(_ v: Double, _ places: Int) -> String {
            String(format: "%.\(places)f", v).replacingOccurrences(of: ".", with: ",")
        }
        func q(_ s: String) -> String { "\"" + s.replacingOccurrences(of: "\"", with: "\"\"") + "\"" }
        var out = "categoria;colocacao;ordem;competidor;cavalo;tempo_final;tempo_bruto_s;tambores;penalidade_s;sem_tempo\n"
        for row in rankingRows(of: eventId, records: records) {
            let p = row.placing
            let r = row.run
            let e = entry(r.entryId)
            let cols = [
                q(r.category ?? ""), p.place.map(String.init) ?? "SAT", "\(p.entryOrder)",
                q(e?.rider ?? r.rider), q(e?.horse ?? r.horse),
                r.noTime ? "SAT" : TimeFormatter.formatElapsed(p.finalNs).replacingOccurrences(of: ".", with: ","),
                dec(Double(r.elapsedRawNs) / 1e9, 3), "\(r.barrelsKnocked)",
                "\(p.penaltyNs / 1_000_000_000)", r.noTime ? "sim" : "não",
            ]
            out += cols.joined(separator: ";") + "\n"
        }
        return out
    }

    // MARK: - persistência
    private func load() {
        guard let data = try? Data(contentsOf: fileURL) else { return }
        let dec = JSONDecoder()
        dec.dateDecodingStrategy = .iso8601
        guard let p = try? dec.decode(Payload.self, from: data) else { return }
        events = p.events
        entries = p.entries
        currentEventId = p.events.contains { $0.id == p.currentEventId } ? p.currentEventId : nil
    }

    private func save() {
        let enc = JSONEncoder()
        enc.dateEncodingStrategy = .iso8601
        enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        let payload = Payload(events: events, entries: entries, currentEventId: currentEventId)
        if let data = try? enc.encode(payload) {
            try? data.write(to: fileURL, options: .atomic)
        }
    }

    /// Divide uma linha de CSV respeitando aspas duplas (planilhas exportam nomes com ponto e vírgula).
    private static func splitCsvLine(_ line: String, separator: Character) -> [String] {
        var out: [String] = []
        var cur = ""
        var quoted = false
        var it = line.makeIterator()
        var pending: Character? = nil
        while let c = pending ?? it.next() {
            pending = nil
            if quoted, c == "\"" {
                if let n = it.next() {
                    if n == "\"" { cur.append("\"") } else { quoted = false; pending = n }
                } else {
                    quoted = false
                }
            } else if c == "\"" {
                quoted = true
            } else if c == separator, !quoted {
                out.append(cur.trimmingCharacters(in: .whitespaces))
                cur = ""
            } else {
                cur.append(c)
            }
        }
        out.append(cur.trimmingCharacters(in: .whitespaces))
        return out
    }
}
